azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+                self.run_op(op_to_release)
+
+    def _do_connect(self, op=None):
+        """
+        Start connecting the transport in response to some operation (which may or may not be an Iot_Connect operation)
+        """
+        # first, we block the queue while we're connecting
+        self._block(op=op)
+
+        # If we're connecting as a side-effect of some other operation (that is not Iot_Connect), then we queueu
+        # that operation to run after the connection is complete.
+        if op.name is not pipeline_ops.names.Iot_Connect:
+            logger.info("{}({}) queueing until connection complete".format(self.name, op.name))
+            self.queue.append(op)
+
+        # callback that gets called after we're connected.
+        def on_connected(op_connect):
+            logger.info("{}({}) connection is complete".format(self.name, op.name))
+            # if we're connecting because some layer above us asked us to connect, we complete that operation
+            # once the connection is established.
+            if op.name is pipeline_ops.names.Iot_Connect:
+                op.error = op_connect.error
+                self.complete_op(op)
+            # and, no matter what, we always unblock the stage when we're done connecting.
+            self._unblock(op=op, error=op_connect.error)
+
+        # call down to the next stage to connect.  We don't use continue_with_translated_op because we have
+        # extra code that needs to run (unblocking the queue) when the connect is complete and
+        # continue_with_translated_op can't do that.
+        logger.info("{}({}) calling down with Iot_Connect".format(self.name, op.name))
+        self.continue_op(pipeline_ops.Iot_Connect(callback=on_connected))
+
+    def on_connected(self):
+        self.connected = True
+        PipelineStage.on_connected(self)
+
+    def on_disconnected(self):
+        self.connected = False
+        PipelineStage.on_disconnected(self)
+
+
+class IotHub_MqttConverter(PipelineStage):
+    """
+    PipelineStage which converts other Iot and IotHub operations into Mqtt operations.  This stage also
+    converts mqtt pipeline events into Iot and IotHub pipeline events.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.subscription_map = {}
+        self.unsubscription_map = {}
+
+    def _run_op(self, op):
+
+        if op.name == pipeline_ops.names.IotHub_SetAuthProviderArgs:
+            # if we get auth provider args from above, we save some, use some to build topic names,
+            # and always pass it down because we know that the Mqtt_Provider stage will also want
+            # to receive these args.
+            self._set_topic_names(device_id=op.device_id, module_id=op.module_id)
+            client_id = op.device_id
+            if op.module_id:
+                client_id += "/" + op.module_id
+            username = op.hostname + "/" + client_id + "/" + "?api-version=2018-06-30"
```

Comment:
>op.hostname + "/" + client_id + "/" + "?api-version=2018-06-30" [](start = 23, length = 63)

I've been thinking lately that string arithmetic should be used sparingly in favor of format strings.

```
username = "{hostname}/{client_id}/?api-version=2018-06-30".format(hostname=op.hostname, client_id=client_id)
```
is just more readable IMO.